rocksdb_admin/admin_handler.cpp

```diff
@@ -888,33 +1092,11 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
     return;
   }
 
-  // Though it is claimed that AWS s3 sdk is a light weight library. However,
-  // we couldn't afford to create a new client for every SST file downloading
-  // request, which is not even on any critical code path. Otherwise, we will
-  // see latency spike when uploading data to production clusters.
-  std::shared_ptr<common::S3Util> local_s3_util;
+
   if (FLAGS_s3_download_limit_mb > 0) {
     request->s3_download_limit_mb = FLAGS_s3_download_limit_mb;
   }
-  {
-    std::lock_guard<std::mutex> guard(s3_util_lock_);
-    if (s3_util_ == nullptr || should_new_s3_client(*s3_util_, request.get())) {
-      // Request with different ratelimit or bucket has to wait for old
-      // requests to drain.
-      while (s3_util_ != nullptr && s3_util_.use_count() > 1) {
-        LOG(INFO) << "There are other downloads happening, wait "
-                  << kS3UtilRecheckSec << " seconds";
-        std::this_thread::sleep_for(std::chrono::seconds(kS3UtilRecheckSec));
-      }
-      // Invoke destructor explicitly to make sure Aws::InitAPI()
-      // and Aps::ShutdownApi() appear in pairs.
-      s3_util_ = nullptr;
-      s3_util_ = common::S3Util::BuildS3Util(request->s3_download_limit_mb,
-                                             request->s3_bucket);
-    }
-    local_s3_util = s3_util_;
-  }
-
+  auto local_s3_util = createLocalS3Util(request->s3_download_limit_mb, request->s3_bucket);
```

Comment:
Could this cause a bottleneck for backup or restoring? If there are quite a few downloads/uploads happening for a particular segment (with saw 'x' as the download limit), another request with 'y' as the download limit could wait for a long time to create the S3 client. Could this cause issues when helix is trying to promote a master?

Reply:
I don't think it would cause a bottleneck. Actually we are using the same logic in the readonly cluster which has more downloading requests/bandwidth than the backup/restore use-case. It wont happen in the master promotion due to this process only happen in offline to slave state transition.

Reply:
This is an issue that we may need to improve in the future. Though slave->master transition doesn't need s3 client, it may be blocked by offline->slave transitions for other DBs. Because they need to compete for helix concurrency, admin threads, etc.

Reply:
yes. we will optimize it in the future when it becomes the performance bottleneck.